mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -0,0 +1,245 @@
+use crate::CardanoNetwork;
+use crate::{entities::Epoch, StdResult};
+
+use super::CardanoCliChainObserver;
+
+/// A runner that uses Pallas library to interact with a Cardano node using N2C Ouroboros mini-protocols
+pub struct PallasChainObserver {
+    socket: PathBuf,
+    network: CardanoNetwork,
+    fallback: Option<super::cli_observer::CardanoCliChainObserver>,
+}
+
+#[derive(Error, Debug)]
+pub enum PallasChainObserverError {
+    #[error("Failed to create new client: {0}")]
+    UnableToCreateClient(anyhow::Error),
+
+    #[error("Acquire failed: {0}")]
+    AcquireFailed(anyhow::Error),
+
+    #[error("Failed to get current era: {0}")]
+    GetCurrentEraFailed(anyhow::Error),
+
+    #[error("Failed to get block epoch number: {0}")]
+    GetBlockEpochNumberFailed(anyhow::Error),
+
+    #[error("Send release failed: {0}")]
+    SendReleaseFailed(anyhow::Error),
+
+    #[error("Send done failed: {0}")]
+    SendDoneFailed(anyhow::Error),
+
+    #[error("Chainsync send done failed: {0}")]
+    ChainSyncSendDoneFailed(anyhow::Error),
+}
+
+impl From<PallasChainObserverError> for ChainObserverError {
+    fn from(err: PallasChainObserverError) -> ChainObserverError {
+        ChainObserverError::PallasObserverError(err)
+    }
+}
+
+impl PallasChainObserver {
+    /// Creates a new PallasObserver while accepting a fallback CliRunner
+    pub fn new_with_fallback(
+        socket: &Path,
+        network: CardanoNetwork,
+        fallback: CardanoCliChainObserver,
+    ) -> Self {
+        Self {
+            socket: socket.to_owned(),
+            network,
+            fallback: Some(fallback),
+        }
+    }
+
+    /// Creates a new PallasObserver
+    pub fn new(socket: &Path, network: CardanoNetwork) -> Self {
+        Self {
+            socket: socket.to_owned(),
+            network,
+            fallback: None,
+        }
+    }
+
+    async fn new_client(&self) -> StdResult<NodeClient> {
+        let magic = self.network.code();
+        let client = NodeClient::connect(&self.socket, magic).await?;
+
+        Ok(client)
+    }
+
+    fn get_fallback(&self) -> StdResult<&CardanoCliChainObserver> {
+        self.fallback.as_ref().ok_or(anyhow!(
+            "Unimplemented and no fallback configured for PallasObserver",
+        ))
+    }
+}
+
+#[async_trait]
+impl ChainObserver for PallasChainObserver {
+    async fn get_current_epoch(&self) -> Result<Option<Epoch>, ChainObserverError> {
+        let mut client = self
+            .new_client()
+            .await
+            .map_err(PallasChainObserverError::UnableToCreateClient)?;
+
+        let statequery = client.statequery();
+
+        statequery
+            .acquire(None)
+            .await
+            .map_err(|err| PallasChainObserverError::AcquireFailed(anyhow!("{:?}", err)))?;
+
+        let era = queries_v16::get_current_era(statequery)
+            .await
+            .map_err(|err| PallasChainObserverError::GetCurrentEraFailed(anyhow!("{:?}", err)))?;
```

Comment:
To illustrate the comment above:
```suggestion
        let era = queries_v16::get_current_era(statequery)
            .await
            .map_err(|err| anyhow!(err))
            .with_context("PallasChainObserver failed retrieving the current era when retrieving the current epoch")?;
```